mesonbuild/backend/ninjabackend.py

```diff
@@ -1102,8 +1102,11 @@ def generate_custom_target(self, target):
         elem.add_dep(deps)
         for d in target.extra_depends:
             # Add a dependency on all the outputs of this target
-            for output in d.get_outputs():
-                elem.add_dep(os.path.join(self.get_target_dir(d), output))
+            if isinstance(d, File):
+                elem.add_dep(d.absolute_path(self.environment.get_source_dir(), self.environment.get_build_dir()))
+            else:
+                for output in d.get_outputs():
+                    elem.add_dep(os.path.join(self.get_target_dir(d), output))
```

Comment:
File objects are supposed to go to custom_target()'s depend_files kwarg. I should probably finish https://github.com/mesonbuild/meson/pull/11061 where I instead deprecate depend_files and merge it with depends. That would solve this issue in a better way IMHO.

Reply:
> Unfortunately, `gnome.generate_gir()` does not exposes `depend_files`, so I think there is no practical solution in your case :/

I think that's a different issue: it's not possible to add `giscanner_pymod` as dependency since it's a `SharedModule`.  In that case, meson correctly errors out since you can pass a `SharedModule` to a `dependencies`.

I think the issue here is different and it can be reproduced easily.

Reply:
The error is in fact in `gnome.generate_gir`, which adds `gir_dep.sources` into `depends`.  I have found a proper way to fix it. I will close this PR